rfcs/0039-minaform-tool-design.md

```diff
@@ -0,0 +1,117 @@
+    - See proposal for how to pre-generate proofs in docker images
+5) Hard to determine the correct docker tag from a given git SHA
+    - Can be scripted to allow for a git tag or git SHA input, in addition to / in place of the docker tag.
+6) Keys and other private info for deployment are not committed/shared with other team members
+    - Addressed by https://github.com/MinaProtocol/coda-automation/issues/472
+7) Incompatible helm/terraform versions
+    - Can be addressed by 1) checking versioning before trying to use these tools and 2) bundling a canonical set of tools into a docker image
+8) Incompatible helm/terraform changes
+    - Addressed by moving the charts to the mina repo and explicitly releasing them. Terraform specifies an individual version of each chart to depend on and will fetch the released chart before deploys.
+9) Helm string processing is inconsistent/hard
+    - Now all charts are passing `helm lint` in the mina repo, and we can add `helm test` in the future as well.
+10) Passing flags to the daemon process directly is difficult to debug/handle/manage
+    -  Move to using daemon.json more and exposing all of these flags through the charts.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+The design revolves around a new bash script, minaform, that can be bundled in a container alongside its dependencies.
+
+### Minaform create-network
+
+The first entrypoint to deploying a new network with minaform is the `create-network <testnet_name>` command which will create a new testnet folder based on the provided template (qa or prod templates to start, but more can be incorporated later), and then generate or download a set of keys + ledger to use with this new network. This templating system addresses Pain Point #1, but still needs cleanup afterwards. See `minaform delete-network <testnet_name>` for more on how to handle that.
+
+With the templated terraform, minaform can now use the `coda-network` tool (aka carey's key service), to look for a set of keys and genesis block in google cloud for this testnet name, and if they do not exist then it will generate all of the neccessary keys for you, as well as the ledger itself.
+
+Flags to the `create-network` command for `--generate-keys`, `--upload-keys`, and `--download-keys` allow you to force regeneration/re-upload of keys for easier testing and quick iteration on the key generation tooling. The default value of `true` for each of these makes it easy to share keys across O(1)  Labs employees who are collaborating on the same network (Pain Point #6).
+
+Additional flags can be provided to the create-network command for changing the quantities of fish, whales, seeds, bots, etc.
+
+The flag `--generate-image` will cause the tool to build a new image based on the provided one, but with the genesis ledger, proofs, and keys baked-in to reduce network startup time at the expense of this one-time step (Pain Point #4).
```

Comment:
This should also not be tied to the creation of a network imo. The genesis proofs need to be regenerated anytime the genesis network configuration changes. It makes more sense to generate/load this from a cache on network deployment instead. @mrmr1993 and I have discussed a system for keying these proofs for caching so that we can avoid regenerating proofs we already have in a cache.